src/basic/copy.c

```diff
@@ -386,14 +386,13 @@ static int fd_copy_directory(
                         continue;
                 }
 
-                if (buf.st_dev != original_device)
-                        continue;
-
                 if (S_ISREG(buf.st_mode))
                         q = fd_copy_regular(dirfd(d), de->d_name, &buf, fdt, de->d_name, override_uid, override_gid, copy_flags);
-                else if (S_ISDIR(buf.st_mode))
+                else if (S_ISDIR(buf.st_mode)) {
+                        if (buf.st_dev != original_device)
+                                continue;
```

Comment:
@rhertzog, I think that the patch fixes the bug and should be updated, but I'd rather wait for someone else to confirm that skipping device nodes and bind-mounted files is not intentional.